pilot/proxy/envoy/discovery.go

```diff
@@ -308,11 +308,15 @@ func (ds *DiscoveryService) Register(container *restful.Container) {
 }
 
 // Run starts the server and blocks
-func (ds *DiscoveryService) Run() {
+func (ds *DiscoveryService) Run(stop chan struct{}) {
 	glog.Infof("Starting discovery service at %v", ds.server.Addr)
 	if err := ds.server.ListenAndServe(); err != nil {
 		glog.Warning(err)
 	}
+
+	// Wait for the stop notification and shutdown the server.
+	<-stop
```

Comment:
`ListenAndServer` blocks forever.

Reply:
ah, thanks for pointing that out.  I've changed it to run this asynchronously.